src/main/scala/com/fulcrumgenomics/util/ReadStructure.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.util
+
+import com.fulcrumgenomics.util.BaseSymbol._
+import com.fulcrumgenomics.util.ReadSegment._
+
+object ReadStructure {
+  private val segmentPattern = s"\\d+[${SymbolChars.mkString}]".r
+  private val readStructurePattern = s"(\\d+[${SymbolChars.mkString}])+".r
+
+  /** Creates a read structure from a string */
+  def apply(readStructure: String): ReadStructure = {
+    if (readStructurePattern.findAllIn(readStructure).mkString.length != readStructure.length) {
+      throw new IllegalArgumentException("Read structure was not valid: " + readStructure)
+    }
+    val segments = segmentPattern.findAllIn(readStructure).toSeq
+    val lengths = segments.map(seg => Integer.parseInt(seg.substring(0, seg.length-1)))
+    new ReadStructure(segs = segments.zipWithIndex.map {
```

Comment:
I think this whole section could be made more readable by doing the splitting from a string once, and then doing a single map.  Something like:
```scala
segmentPattern.findAllIn(readStructure)
  .map (str => str.splitAt(str.length-1))
  .zipWithIndex
  .map { case ((len, kind), idx) => (len.toInt, kind.charAt(0), idx) }
  .map { 
    case (0, _, _) => throw new IllegalArgumentException("Read structure had a segment with zero length: " + seg)
    case (n, TemplateSymbol.symbol, idx) => ....
    ....
  }
```